### PR TITLE
✨ Adds placeholder transition from blurred to amp-video posters

### PR DESCRIPTION
**Pull request by addieachan**
- Addresses issues #17873 and #15146
 -Follow up PR to #17132

### CODE REVIEW

**Comment by addieachan** [comment]
PTAL @choumx , @gmajoulet

File: extensions/amp-video/0.1/amp-video.js [hunk 1]
```diff
@@ -618,6 +619,19 @@ class AmpVideo extends AMP.BaseElement {
...
   }
 
+  /**
+    * @override
+    **/
+  firstLayoutCompleted() {
+    const placeholder = this.getPlaceholder();
+    if (placeholder && placeholder.classList.contains('i-amphtml-blur') &&
+      isExperimentOn(this.win, 'blurry-placeholder')) {
+      setImportantStyles(placeholder, {'opacity': 0});
```
Comment (choumx) [inline extensions/amp-video/0.1/amp-video.js]: Add a comment explaining that this will trigger a CSS fade out animation for the placeholder.

File: extensions/amp-video/0.1/amp-video.js [hunk 2]
```diff
@@ -618,6 +619,19 @@ class AmpVideo extends AMP.BaseElement {
     return ranges;
   }
 
+  /**
+    * @override
+    **/
+  firstLayoutCompleted() {
```
Comment (choumx) [inline extensions/amp-video/0.1/amp-video.js]: Add a comment that this is called when the video is loaded.

Comment (choumx) [inline extensions/amp-video/0.1/amp-video.js]: What if the poster image is loaded before the video is loaded?

**Comment by addieachan** [comment]
@choumx @gmajoulet @alanorozco PTAL

File: extensions/amp-video/0.1/amp-video.js [hunk 3]
```diff
@@ -198,6 +198,10 @@ class AmpVideo extends AMP.BaseElement {
     this.createPosterForAndroidBug_();
     element.appendChild(this.video_);
 
+    // Fade out blur on poster load.
+    const posterEl = element.getElementsByTagName('i-amphtml-poster');
+    posterEl.onload = this.fadeOutBlur_();
```
Comment (choumx) [inline extensions/amp-video/0.1/amp-video.js]: This won't work if you invoke `fadeOutBlur_()` immediately.

File: extensions/amp-video/0.1/amp-video.js [hunk 4]
```diff
@@ -198,6 +198,10 @@ class AmpVideo extends AMP.BaseElement {
     this.createPosterForAndroidBug_();
     element.appendChild(this.video_);
 
+    // Fade out blur on poster load.
+    const posterEl = element.getElementsByTagName('i-amphtml-poster');
```
Comment (choumx) [inline extensions/amp-video/0.1/amp-video.js]: This only exists on Android devices. `poster` is a built-in attribute on the `<video>` element. 

See if you can figure out how to detect poster loading on videos.

File: extensions/amp-video/0.1/amp-video.js [hunk 5]
```diff
@@ -622,13 +631,56 @@ class AmpVideo extends AMP.BaseElement {
...
+    }
+  }
+
+  /**
+   * Adds an onloading event for the actual poster element for the case where
+   * the poster has a blurry image placeholder and loads before the video
+   * element.
+   * @param {*function()} opacityChangeFunc
+   * @private
+   */
```
Comment (addieachan) [inline extensions/amp-video/0.1/amp-video.js]: Where I got this logic:
https://stackoverflow.com/questions/18571424/how-to-add-event-listener-to-html5-video-poster-image-load-event
https://stackoverflow.com/questions/33799763/event-on-video-poster-loaded?noredirect=1&lq=1

**Comment by addieachan** [comment]
PTAL @choumx @alanorozco

**Comment by addieachan** [comment]
Accidentally closed

File: extensions/amp-video/0.1/amp-video.js [hunk 6]
```diff
@@ -622,13 +631,56 @@ class AmpVideo extends AMP.BaseElement {
...
   }
 
+  /**
+    * Called when the video is first loaded.
+    * @override
+  */
+  firstLayoutCompleted() {
+    if (this.getPlaceholder() == null) {
+      this.togglePlaceholder(false);
+    }
```
Comment (choumx) [inline extensions/amp-video/0.1/amp-video.js]: I'm reading this as "if there's no placeholder, hide it".

File: build-system/tasks/bundle-size.js [hunk 7]
```diff
@@ -24,6 +24,7 @@ const {getStdout} = require('../exec');
 const runtimeFile = './dist/v0.js';
 const maxSize = '81.4KB'; // Only use 0.1KB of precision i.e no hundredths digit.
 
+
```
Comment (choumx) [inline build-system/tasks/bundle-size.js]: Revert this change.

File: extensions/amp-video/0.1/amp-video.js [hunk 8]
```diff
@@ -197,6 +201,11 @@ class AmpVideo extends AMP.BaseElement {
     this.createPosterForAndroidBug_();
     element.appendChild(this.video_);
 
+    const posterFunc = () => {
+      this.fadeOutBlur_();
+    };
```
Comment (choumx) [inline extensions/amp-video/0.1/amp-video.js]: Inline this.

File: extensions/amp-video/0.1/amp-video.js [hunk 9]
```diff
@@ -622,13 +631,56 @@ class AmpVideo extends AMP.BaseElement {
...
   getUrlService_() {
     return Services.urlForDoc(this.element);
   }
+
+  /**
+    * Fades out a blurry placeholder if one currently exists.
+    * @return {boolean} whether or not there was a blurry placeholder that was
+    * faded.
+    */
+  fadeOutBlur_() {
```
Comment (choumx) [inline extensions/amp-video/0.1/amp-video.js]: `hideBlurryPlaceholder_()`

File: extensions/amp-video/0.1/amp-video.js [hunk 10]
```diff
@@ -622,13 +631,56 @@ class AmpVideo extends AMP.BaseElement {
...
    * @private
    */
   getUrlService_() {
     return Services.urlForDoc(this.element);
   }
+
+  /**
+    * Fades out a blurry placeholder if one currently exists.
+    * @return {boolean} whether or not there was a blurry placeholder that was
+    * faded.
```
Comment (choumx) [inline extensions/amp-video/0.1/amp-video.js]: This function doesn't appear to return anything.

File: extensions/amp-video/0.1/amp-video.js [hunk 11]
```diff
@@ -622,13 +631,56 @@ class AmpVideo extends AMP.BaseElement {
...
+  }
+
+  /**
+   * Adds an onloading event for the actual poster element for the case where
+   * the poster has a blurry image placeholder and loads before the video
+   * element.
+   * @param {function()} opacityChangeFunc
+   * @private
+   */
+  getPosterFromVideo_(opacityChangeFunc) {
```
Comment (choumx) [inline extensions/amp-video/0.1/amp-video.js]: `onPosterLoaded_(callback)`

File: extensions/amp-video/0.1/amp-video.js [hunk 12]
```diff
@@ -622,13 +631,56 @@ class AmpVideo extends AMP.BaseElement {
...
+    const poster = this.video_.getAttribute('poster');
+    if (poster != null) {
+      this.posterImg_ = new Image();
+      this.posterImg_.src = poster;
```
Comment (choumx) [inline extensions/amp-video/0.1/amp-video.js]: Do any browsers fetch the poster twice here instead once?

File: extensions/amp-video/0.1/amp-video.js [hunk 13]
```diff
@@ -622,13 +631,56 @@ class AmpVideo extends AMP.BaseElement {
...
+    * faded.
+    */
+  fadeOutBlur_() {
+    const placeholder = this.getPlaceholder();
+    // checks for the existance of a visible blurry placeholder
+    if (placeholder && placeholder.classList.contains('i-amphtml-blur') &&
+      isExperimentOn(this.win, 'blurry-placeholder')) {
+      setImportantStyles(placeholder, {'opacity': 0.0});
+    } else {
+      this.togglePlaceholder(false);
```
Comment (choumx) [inline extensions/amp-video/0.1/amp-video.js]: We should avoid calling this if `placeholder` is null.

File: extensions/amp-video/0.1/amp-video.js (hunk 6 above)
Comment (addieachan) [reply extensions/amp-video/0.1/amp-video.js]: Sorry forgot to push the commit that deleted this.

File: extensions/amp-video/0.1/amp-video.js (hunk 12 above)
Comment (addieachan) [reply extensions/amp-video/0.1/amp-video.js]: I'm researching how to check this, and will let you know ASAP!

**Comment by addieachan** [comment]
@choumx If I addressed your last comments properly then can you please help me merge?

File: extensions/amp-video/0.1/amp-video.js [hunk 14]
```diff
@@ -622,13 +628,58 @@ class AmpVideo extends AMP.BaseElement {
     return ranges;
   }
 
+  /**
+   * Called when video is first loaded.
+   * @override
+   */
+  firstLayoutCompleted() {
+    if (this.getPlaceholder() && !this.hideBlurryPlaceholder_()) {
```
Comment (choumx) [inline extensions/amp-video/0.1/amp-video.js]: Let's remove `this.getPlaceholder()`. My old comment is obsolete now that it's moved outside of `hideBlurryPlaceholder_` and I looked at `togglePlaceholder` and it also does this check.

File: extensions/amp-video/0.1/amp-video.js [hunk 15]
```diff
@@ -622,13 +628,58 @@ class AmpVideo extends AMP.BaseElement {
...
    * @private
    */
   getUrlService_() {
     return Services.urlForDoc(this.element);
   }
+
+  /**
+  * Fades out a blurry placeholder if one currently exists.
+  * @return {boolean} if there was a blurred image placeholder that was hidden.
+  */
```
Comment (choumx) [inline extensions/amp-video/0.1/amp-video.js]: Nit: Fix indentation.

File: extensions/amp-video/0.1/test/test-amp-video.js [hunk 16]
```diff
@@ -479,6 +479,100 @@ describes.realWin('amp-video', {
...
+      impl.firstLayoutCompleted();
+      el = impl.element;
+      img = el.firstChild;
+      expect(impl.togglePlaceholder).to.not.be.called;
+    });
+
+    it('should fade out the blurry image placeholder on video load', () => {
+      const impl = getVideoWithBlur(true, true);
+      impl.buildCallback();
+      impl.firstLayoutCompleted();
```
Comment (choumx) [inline extensions/amp-video/0.1/test/test-amp-video.js]: Can we call `layoutCallback` before this? It's confusing that the layout completes without being called. :)
